File: pganalyze-collector.rb
#!/usr/bin/env ruby1.9.1

require 'rubygems'
require 'json'

require 'net/http'
require 'getoptlong'
require 'yaml'
require 'logger'

require 'pp'

API_URL = 'http://pganalyze.com/queries'
RESET_STATS = true

MYNAME = 'pganalyze-collector'
VERSION = '0.0.1'

opts = GetoptLong.new(
	[ '--help', '-h', GetoptLong::NO_ARGUMENT ],
	[ '--verbose', '-v', GetoptLong::NO_ARGUMENT ],
	[ '--version', '-V', GetoptLong::NO_ARGUMENT ],
	[ '--config', GetoptLong::REQUIRED_ARGUMENT ],
	[ '--generate-config', GetoptLong::NO_ARGUMENT ],
	[ '--cron', '--quiet', GetoptLong::NO_ARGUMENT ]
)

$help = <<"BE_DRAGONS"

#{MYNAME} #{VERSION}

-h, --help
	This help

-v, --verbose
	Print verbose debug information

-V, --version
	Print version information

--config CONFIGFILE
	Specifiy alternative path for config file.
	Defaults: $HOME/.pganalyze_collector.conf, /etc/pganalyze/collector.conf

--generate-config
	Writes a default configuration file to $HOME/.pganalyze_collector.conf
	unless specified otherwise with --config

--cron, --quiet
	Suppress all non-warning output during normal operation

BE_DRAGONS

$configfile = [ ENV['HOME'] + '/.pganalyze_collector.conf', '/etc/pganalyze/collector.conf' ] 

$logger = Logger.new(STDOUT)
$logger.level = Logger::INFO

opts.each do |opt, arg|
	case opt
	when '--help'
		puts $help
		exit
	when '--verbose'
		$logger.level = Logger::DEBUG
	when '--version'
		puts "#{MYNAME} #{VERSION}"
		exit
	when '--config'
		$configfile = [arg]
	when '--cron'
		$quiet = TRUE
	when '--generate-config'
		$generate_config = TRUE
	end
end

class PSQL
	def initialize(*args)
		(@host, @port, @username, @password, @dbname, @psql) =  args[0].values_at(:host, :port, :username, :password, :dbname, :psql)
		@host ||= 'localhost'
		@port ||= 5432
		@psql ||= find_psql
		raise "Please specify a database" unless @dbname
		raise "Please specify path to psql binary" unless @psql
		
		# Setting up ENV for psql
		ENV['PGUSER'] = @username
		ENV['PGPASSWORD'] = @password
		ENV['PGHOST'] = @host
		ENV['PGPORT'] = @port.to_s
		ENV['PGDATABASE'] = @dbname
	end

	def exec(query, should_raise = false)
		# FIXME: ruby1.8 popen only supports command strings, not arrays

		$logger.debug "Running query: #{query}"

		err_rd, err_wr = IO.pipe
		cmd = [@psql, "-F\u2764", '--no-align', '--no-password', '--no-psqlrc', "-c", query, :err => err_wr]
		lines = []

		IO.popen(cmd, :err => err_wr) { |f| lines = f.readlines }

		begin
			stderr = err_rd.read_nonblock(16384).split(/\n/)
		rescue Errno::EAGAIN
			# Fall through if stderr is empty
		end

		if stderr or $?.exitstatus != 0
			if should_raise
				raise RuntimeError, stderr
			end
			$logger.error "Got an error during query execution, exitstatus: #{$?.exitstatus}:"
			stderr.each { |l| $logger.error l }
			exit
		end

		# Drop number of rows
		lines.pop
		# Fetch column headers
		columns = lines.shift.strip.split(/\u2764/)

		resultset = []
		lines.each { |line|
		  values = line.strip.split(/\u2764/)
			resultset.push(Hash[columns.zip(values)])
		}

		return resultset
	end

	def ping
		$logger.debug "Pinging database"
		exec('SELECT 1')
	end

	private
	def find_psql
		cmd = 'psql'
		ENV['PATH'].split(File::PATH_SEPARATOR).each do |path|
			test = "#{path}/#{cmd}"
			return test if File.executable? test
		end
		return nil
	end
end


def post_data_to_web(queries)
	to_post = {}
	to_post['data'] = {'queries' => queries}.to_json
	to_post['api_key'] = $api_key
	to_post['collected_at'] = Time.now.to_i

	begin
		res = Net::HTTP.post_form(URI.parse(API_URL), to_post)
		return res.code == '200'
	rescue => e
		$logger.error "Failed to post data to service: #{e.message}"
	end
end

def fetch_queries
	both_fields = ["userid", "dbid",
		"calls", "rows", "total_time",
		"shared_blks_hit", "shared_blks_read", "shared_blks_written",
		"local_blks_hit", "local_blks_written",
		"temp_blks_read", "temp_blks_written"]

	query_fields = ["plan_ids", "calls_per_plan", "avg_time_per_plan",
		"time_variance", "time_stddev"] + both_fields

	plan_fields = ["planid", "had_our_search_path", "from_our_database",
		"query_valid", "last_startup_cost", "last_total_cost"] + both_fields

	query = "SET pg_stat_plans.explain_format TO JSON;"
	query += "SELECT replace(pg_stat_plans_explain(p.planid, p.userid, p.dbid), chr(10), ' ') AS p_explain"
	query += ", replace(pq.normalized_query, chr(10), ' ') AS pq_normalized_query"
	query += ", replace(p.query, chr(10), ' ') AS p_query"
	query += ", " + query_fields.map {|f| "pq.#{f} AS pq_#{f}"}.join(", ")
	query += ", " + plan_fields.map {|f| "p.#{f} AS p_#{f}"}.join(", ")
	query += " FROM pg_stat_plans p"
	query += " LEFT JOIN pg_stat_plans_queries pq ON p.planid = ANY (pq.plan_ids)"
	# EXPLAIN, COPY and SET commands cannot be explained
	query += " WHERE p.query !~* '^\\s*(EXPLAIN|COPY|SET)'"
	# Plans in pg_catalog cannot be explained
	query += " AND p.query !~* '\\spg_catalog\\.'"
	# We don't want our stuff in the statistics
	query += " AND p.query !~* '\\spg_stat_plans\\s'"
	# Remove all plans which we can't explain
	query += " AND p.from_our_database = TRUE AND p.query_valid = TRUE"
	query += " AND p.planid = ANY (pq.plan_ids);"

	queries = {}
	$db.exec(query).each do |row|
		query = {}; row.select {|k,| k[/^pq_/] }.each {|k,v| query[k.gsub("pq_", "")] = v }
		key = query['normalized_query']
		queries[key] ||= query

		plan = {}; row.select {|k,| k[/^p_/] }.each {|k,v| plan[k.gsub("p_", "")] = v }
		queries[key]['plans'] ||= []
		queries[key]['plans'] << plan
	end
	queries.values
end

def check_database
	$db = PSQL.new(:host => $db_host,
		       :port => $db_port,
		       :username => $db_username,
		       :password => $db_password,
		       :dbname => $db_name)

	unless $db.ping
		$logger.error "DB is not alive"
		exit
	end
	
	unless $db.exec('SHOW is_superuser')[0]['is_superuser'] == 'on'
		$logger.error "User #{$db_username} isn't a superuser"
		exit
	end

	unless $db.exec('SHOW server_version_num')[0]['server_version_num'].to_i >= 90100
		$logger.error "You must be running PostgreSQL 9.1 or newer"
		exit
	end

	begin
		unless $db.exec("SELECT COUNT(*) as foo FROM pg_extension WHERE extname='pg_stat_plans'", true)[0]['foo'] == '1'
			$logger.error "Extension pg_stat_plans isn't installed"
			exit
		end
	rescue
		$logger.error "Table pg_extension doesn't exist - this shouldn't happen!"
		exit
	end
end

def read_config
	# Return the first readable config file

	configfile = nil
	$configfile.each { |f|
		stat = nil
		begin 
			stat = File.stat(f)
		rescue Errno::ENOENT
			$logger.debug "#{f} doesn't exist"
			next
		end

		unless stat.file?
			$logger.debug "#{f} isn't a regular file"
			next
		end

		unless stat.readable?
			$logger.debug "#{f} isn't readable"
			next
		end
		configfile = f
		break
	}

	unless configfile
		$logger.error "Couldn't find a config file, perhaps create one with --generate-config?"
		exit
	end

	configdump = {}

	begin
		File.open(configfile) { |yf| configdump = YAML::load(yf) }
	rescue ArgumentError
		$logger.error "Failure while parsing #{configfile}, please fix or create a new one with --generate-config" 
		exit
	end

	$logger.debug "read config from #{$configfile[0]}"
	configdump.each { |k, v|
		$logger.debug "#{k} => #{v}"
	}

	$db_host = configdump['db_host']
	$db_port = configdump['db_port']
	$db_username = configdump['db_username']
	$db_password = configdump['db_password']
	$db_name = configdump['db_name']
	$api_key = configdump['api_key']
	$psql_binary = configdump['psql_binary']

	unless $db_name && $api_key
		$logger.error "Missing database name and/or api key in configfile #{configfile}, perhaps create one with --generate-config?"
		exit
	end
end

def write_config

	config = <<'EOF'
api_key: fill_me_in
db_name: fill_me_in
#db_username:
#db_password:
#db_host: localhost
#db_port: 5432
#psql_binary: /autodetected/from/$PATH
EOF


	cf = $configfile[0]

	begin
		File.open(cf, File::WRONLY|File::CREAT|File::EXCL) { |f|
			f.write(config)
		}
	rescue => e
		$logger.error "Failed to write configfile: #{e.message}"
		exit
	end
	$logger.info "Wrote standard configuration to #{cf}, please edit it and then run the script again"
end

# Start of main

if $print_help
	print_help
	exit
end

if $generate_config
	write_config
	exit
end

unless $db_name && $api_key
	read_config
end

check_database

queries = fetch_queries

# FIXME: Verbose error reporting for wrong API key/broken data/etc?
if post_data_to_web(queries)
  unless $quiet
	  $logger.info "Submitted successfully"
  end
  if RESET_STATS
    $logger.debug "Resetting stats!"
    $db.exec("SELECT pg_stat_plans_reset()")
  end
else
  $logger.error "Rejected by server"
end

File: pganalyze-collector.py
#!/usr/bin/env python

# Set up vendor include path
import sys
sys.path.insert(1, sys.path[0] + '/vendor/')

import os
import time
import calendar
import datetime
import re
import json
import urllib
import logging
from optparse import OptionParser

compressor_lib = None

try:
    import zlib as compressor
    compressor_lib = 'zlib'
except Exception as e:
    pass

from pgacollector.PostgresInformation import PostgresInformation
from pgacollector.PgStatStatements import PgStatStatements
from pgacollector.SystemInformation import SystemInformation
from pgacollector.DB import DB
from pgacollector.Configuration import Configuration

MYNAME = 'pganalyze-collector'
VERSION = '0.8.1'
API_URL = 'https://api.pganalyze.com/v1/snapshots'
dbconf = {}

def setup_database():
    return DB(querymarker=MYNAME, host=dbconf['host'], port=dbconf['port'], username=dbconf['username'],
              password=dbconf['password'], dbname=dbconf['dbname'])

def is_remote_system():
    global dbconf
    is_awshost = dbconf['host'] != None and re.search('amazonaws.com$', dbconf['host']) != None
    return is_awshost or SystemInformation().on_heroku

def parse_options(print_help=False):
    parser = OptionParser(usage="%s [options]" % MYNAME, version="%s %s" % (MYNAME, VERSION))

    parser.add_option('-v', '--verbose', action='store_true', dest='verbose',
                      help='Print verbose debug information')
    parser.add_option('--config', action='store', type='string', dest='configfile',
                      default='$HOME/.pganalyze_collector.conf, /etc/pganalyze_collector.conf',
                      help='Specifiy alternative path for config file. Defaults: %default')
    parser.add_option('--generate-config', action='store_true', dest='generate_config',
                      help='Writes a default configuration file to $HOME/.pganalyze_collector.conf unless specified otherwise with --config')
    parser.add_option('--api-key', action='store', type='string', dest='apikey',
                      help='Use specified API key when writing a fresh config file')
    parser.add_option('--cron', '-q', action='store_true', dest='quiet',
                      help='Suppress all non-warning output during normal operation')
    parser.add_option('--dry-run', '-d', action='store_true', dest='dryrun',
                      help='Print JSON data that would get sent to web service and exit afterwards.')
    parser.add_option('--no-postgres-settings', action='store_false', dest='collect_postgres_settings',
                      default=True,
                      help='Don\'t collect Postgres configuration settings')
    parser.add_option('--no-postgres-locks', action='store_false', dest='collect_postgres_locks',
                      default=True,
                      help='Don\'t collect Postgres lock information')
    parser.add_option('--no-postgres-functions', action='store_false', dest='collect_postgres_functions',
                      default=True,
                      help='Don\'t collect Postgres function/procedure information')
    parser.add_option('--no-postgres-queries', action='store_false', dest='collect_postgres_queries',
                      default=True,
                      help='Don\'t collect Postgres queries information')
    parser.add_option('--no-postgres-bloat', action='store_false', dest='collect_postgres_bloat',
                      default=True,
                      help='Don\'t collect Postgres table/index bloat statistics')
    parser.add_option('--no-postgres-views', action='store_false', dest='collect_postgres_views',
                      default=True,
                      help='Don\'t collect Postgres view/materialized view information')
    parser.add_option('--no-system-information', action='store_false', dest='systeminformation',
                      default=True,
                      help='Don\'t collect OS level performance data')
    parser.add_option('--no-reset', '-n', action='store_true', dest='_dummy_noreset',
		              help='Dummy option, no-reset is required default since 0.7')
    parser.add_option('--no-compression', action='store_false', dest='compression_enabled',
                      default=True,
                      help='Disable gzip compression for statistics data sent')

    if print_help:
        parser.print_help()
        return

    (options, args) = parser.parse_args()
    options = options.__dict__
    options['configfile'] = re.split(',\s+', options['configfile'].replace('$HOME', os.environ['HOME']))
    options['api_url'] = API_URL

    return options


def configure_logger():
    logtemp = logging.getLogger(MYNAME)

    loglevel = logging.DEBUG if option['verbose'] else logging.INFO
    logformat = '%(levelname)s - %(asctime)s %(message)s'

    logargs = {
        'format': logformat,
        'level': loglevel,
    }
    logging.basicConfig(**logargs)

    return logtemp


def fetch_system_information():
    SI = SystemInformation(db)
    info = {}

    info['os'] = SI.os()
    info['cpu'] = SI.cpu()
    info['scheduler'] = SI.scheduler()
    info['storage'] = SI.storage()
    info['memory'] = SI.memory()

    return info


def fetch_postgres_information():
    """
    Fetches information about the Postgres installation

    Returns a groomed version of all info ready for posting to the web
"""
    PI = PostgresInformation(db)

    info = {}
    schema = {}

    table_bloat_stats = {}
    index_bloat_stats = {}

    if option['collect_postgres_bloat']:
        for row in PI.table_bloat():
            table_bloat_stats[row['oid']] = row['wasted_bytes']

        for row in PI.index_bloat():
            index_bloat_stats[row['index_oid']] = row['wasted_bytes']

    for row in PI.relations(option['collect_postgres_views']):
        oid = row.pop('oid')
        schema[oid] = dict((k, row[k]) for k in ('schema_name', 'table_name', 'relation_type'))
        schema[oid]['stats'] = dict((k, row[k]) for k in set(row.keys()) - set(['relid', 'relname', 'schema_name', 'schemaname', 'table_name', 'relation_type']))
        schema[oid]['stats']['wasted_bytes'] = table_bloat_stats.get(oid)
        schema[oid]['columns'] = []
        schema[oid]['indices'] = []
        schema[oid]['constraints'] = []

    if option['collect_postgres_views']:
        for row in PI.view_definitions():
            schema[row['oid']]['view_definition'] = row['view_definition']

    for row in PI.columns(option['collect_postgres_views']):
        oid = row.pop('oid')
        schema[oid]['columns'].append(row)

    for row in PI.indexes(option['collect_postgres_views']):
        oid = row.pop('oid')
        row['wasted_bytes'] = index_bloat_stats.get(row.pop('index_oid'))
        schema[oid]['indices'].append(row)

    for row in PI.constraints():
        oid = row.pop('oid')
        schema[oid]['constraints'].append(row)

    # Populate result dictionary
    info['schema']   = schema.values()
    info['version']  = PI.version()
    info['server']   = PI.server_stats()
    info['database'] = PI.db_stats()
    info['bgwriter'] = PI.bgwriter_stats()
    info['backends'] = PI.backends()
    info['replication']           = PI.replication()
    info['replication_conflicts'] = PI.replication_conflicts()

    if option['collect_postgres_functions']:
        info['functions'] = PI.functions()

    if option['collect_postgres_settings']:
        info['settings'] = PI.settings()

    if option['collect_postgres_locks']:
        info['locks']    = PI.locks()

    return info


class DatetimeEncoder(json.JSONEncoder):

    def default(self, obj):
        if isinstance(obj, datetime.datetime):
            return str(obj)

        return json.JSONEncoder.default(self, obj)


def post_data_to_web(data):
    data = json.dumps(data, cls=DatetimeEncoder)

    to_post = {}

    if option['compression_enabled'] and compressor_lib:
        logger.debug("Compressing data using %s", compressor_lib)
        to_post['data'] = compressor.compress(data)
        to_post['data_compressor'] = compressor_lib
    else:
        to_post['data'] = data

    to_post['api_key'] = dbconf['api_key']
    to_post['collected_at'] = calendar.timegm(time.gmtime())
    to_post['submitter'] = "%s %s" % (MYNAME, VERSION)
    to_post['system_information'] = option['systeminformation']
    to_post['query_source'] = option.get('query_source')
    to_post['no_reset'] = True

    if option['dryrun']:
        logger.info("Dumping data that would get posted")

        to_post['data'] = json.loads(data)
        print(json.dumps(to_post, sort_keys=True, indent=4, separators=(',', ': '), cls=DatetimeEncoder))

        logger.info("Exiting.")
        sys.exit(0)

    num_tries = 0
    while True:
        try:
            # FIXME: urllib doesn't do any SSL verification
            res = urllib.urlopen(dbconf['api_url'], urllib.urlencode(to_post))
            message = res.read()
            code = res.getcode()
        except IOError as e:
            message = str(e)
            code = 'exception'

        num_tries += 1
        if code == 200 or message == 'ERROR: Invalid API key' or num_tries >= 3:
            return message,code
        if not option['quiet']:
            logger.info("Got %s while posting data: %s, sleeping 60 seconds then trying again" % (code, message))
        time.sleep(60)


def fetch_query_information():
    query = "SELECT extname FROM pg_extension"

    extensions = map(lambda q: q['extname'], db.run_query(query))
    if 'pg_stat_statements' in extensions:
        logger.debug("Found pg_stat_statements, using it for query information")
        return ['pg_stat_statements', PgStatStatements(db).fetch_queries()]
    else:
        logger.debug("Trying to enable pg_stat_statements...")
        db.run_query("CREATE EXTENSION IF NOT EXISTS pg_stat_statements", commit = True)
        return ['pg_stat_statements', PgStatStatements(db).fetch_queries()]


def main():
    global option, logger, dbconf, db

    option = parse_options()
    logger = configure_logger()

    c = Configuration(option)

    if option['generate_config']:
        c.write()
        sys.exit(0)

    dbconf = c.read()
    db     = setup_database()

    if db.version_numeric < 90200:
        logger.error("To use the collector you must have at least Postgres 9.2 or newer")
        sys.exit(1)

    if is_remote_system():
        option['systeminformation'] = False

    data = {}
    if option['collect_postgres_queries']:
        (option['query_source'], data['queries']) = fetch_query_information()

    if option['systeminformation']:
        data['system'] = fetch_system_information()

    data['postgres'] = fetch_postgres_information()

    (output, code) = post_data_to_web(data)
    if code == 200:
        if not option['quiet']:
            logger.info("Submitted successfully")
    else:
        logger.error("Rejected by server: %s" % output)


if __name__ == '__main__':
    main()
